File: akka-service-locator/javadsl/src/main/java/com/lightbend/lagom/javadsl/akka/discovery/AkkaDiscoveryServiceLocator.java
/*
 * Copyright (C) 2016-2019 Lightbend Inc. <https://www.lightbend.com>
 */
package com.lightbend.lagom.javadsl.akka.discovery;

import akka.actor.ActorSystem;
import akka.discovery.Discovery;
import com.lightbend.lagom.internal.client.AkkaDiscoveryHelper;
import com.lightbend.lagom.javadsl.api.Descriptor;
import com.lightbend.lagom.javadsl.client.CircuitBreakersPanel;
import com.lightbend.lagom.javadsl.client.CircuitBreakingServiceLocator;
import scala.collection.JavaConverters;
import scala.compat.java8.FutureConverters;
import scala.compat.java8.OptionConverters;

import javax.inject.Inject;
import java.net.URI;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.CompletionStage;

/**
 * Akka Discovery based implementation of the {@link com.lightbend.lagom.javadsl.api.ServiceLocator}.
 */
public class AkkaDiscoveryServiceLocator extends CircuitBreakingServiceLocator {
    private final AkkaDiscoveryHelper helper;

    @Inject
    public AkkaDiscoveryServiceLocator(CircuitBreakersPanel circuitBreakersPanel, ActorSystem actorSystem) {
        super(circuitBreakersPanel);
        this.helper = new AkkaDiscoveryHelper(
                actorSystem.settings().config().getConfig("lagom.akka.discovery"),
                Discovery.get(actorSystem).discovery(), actorSystem.dispatcher());
    }

    @Override
    public CompletionStage<List<URI>> locateAll(String name, Descriptor.Call<?, ?> serviceCall) {
        return FutureConverters.toJava(helper.locateAll(name))
            .thenApply(xs -> JavaConverters.seqAsJavaListConverter(xs).asJava());
    }

    @Override
    public CompletionStage<Optional<URI>> locate(String name, Descriptor.Call<?, ?> serviceCall) {
        return FutureConverters.toJava(helper.locate(name))
            .thenApply(OptionConverters::toJava);
    }
}

File: akka-service-locator/scaladsl/src/main/scala/com/lightbend/lagom/scaladsl/akka/discovery/AkkaDiscoveryServiceLocator.scala
/*
 * Copyright (C) Lightbend Inc. <https://www.lightbend.com>
 */

package com.lightbend.lagom.scaladsl.akka.discovery

import java.net.URI

import akka.actor.ActorSystem
import akka.discovery.Discovery
import com.lightbend.lagom.internal.client.AkkaDiscoveryHelper
import com.lightbend.lagom.scaladsl.api.Descriptor
import com.lightbend.lagom.scaladsl.api.ServiceLocator
import com.lightbend.lagom.scaladsl.client.CircuitBreakersPanel
import com.lightbend.lagom.scaladsl.client.CircuitBreakingServiceLocator

import scala.concurrent.ExecutionContext
import scala.concurrent.Future

/**
 * Akka discovery based implementation of the [[ServiceLocator]].
 */
class AkkaDiscoveryServiceLocator(circuitBreakers: CircuitBreakersPanel, actorSystem: ActorSystem)(
    implicit
    ec: ExecutionContext
) extends CircuitBreakingServiceLocator(circuitBreakers) {

  private val helper: AkkaDiscoveryHelper = new AkkaDiscoveryHelper(
    actorSystem.settings.config.getConfig("lagom.akka.discovery"),
    Discovery(actorSystem).discovery
  )

  override def locate(name: String, serviceCall: Descriptor.Call[_, _]): Future[Option[URI]] =
    helper.locate(name)

  override def locateAll(name: String, serviceCall: Descriptor.Call[_, _]): Future[List[URI]] =
    helper.locateAll(name).map(_.toList)
}
